packages/rum/src/domain/segmentCollection/deflateWorkerSingleton.ts

```diff
@@ -1,32 +1,97 @@
 import { addErrorToMonitoringBatch, display, includes, monitor } from '@datadog/browser-core'
 import { createDeflateWorker, DeflateWorker } from './deflateWorker'
 
-let workerSingleton: DeflateWorker
-
-export function getDeflateWorkerSingleton() {
-  if (!workerSingleton) {
-    try {
-      workerSingleton = createDeflateWorker()
-    } catch (error) {
-      display.error('Session Replay recording failed to start: an error occurred while creating the Worker:', error)
-      if (includes(error.message, 'Content Security Policy')) {
-        display.error(
-          'Please make sure CSP is correctly configured ' +
-            'https://docs.datadoghq.com/real_user_monitoring/faq/content_security_policy'
-        )
-      } else {
-        addErrorToMonitoringBatch(error)
-      }
-      return
+const enum DeflateWorkerSingletonStatus {
+  Nil,
+  Loading,
+  Error,
+  Ready,
+}
+
+type DeflateWorkerSingletonState =
+  | {
+      status: DeflateWorkerSingletonStatus.Nil
+    }
+  | {
+      status: DeflateWorkerSingletonStatus.Loading
+      callbacks: Array<(worker?: DeflateWorker) => void>
+    }
+  | {
+      status: DeflateWorkerSingletonStatus.Error
+    }
+  | {
+      status: DeflateWorkerSingletonStatus.Ready
+      worker: DeflateWorker
     }
-    workerSingleton.addEventListener(
+
+export let state: DeflateWorkerSingletonState = { status: DeflateWorkerSingletonStatus.Nil }
+
+export function loadDeflateWorkerSingleton(
+  callback: (worker?: DeflateWorker) => void,
+  createDeflateWorkerImpl = createDeflateWorker
+) {
+  switch (state.status) {
+    case DeflateWorkerSingletonStatus.Nil:
+      state = { status: DeflateWorkerSingletonStatus.Loading, callbacks: [callback] }
+      startDeflateWorkerSingleton(createDeflateWorkerImpl)
+      break
+    case DeflateWorkerSingletonStatus.Loading:
+      state.callbacks.push(callback)
+      break
+    case DeflateWorkerSingletonStatus.Error:
+      callback()
+      break
+    case DeflateWorkerSingletonStatus.Ready:
+      callback(state.worker)
+      break
+  }
+}
+
+export function resetDeflateWorkerSingletonState() {
+  state = { status: DeflateWorkerSingletonStatus.Nil }
+}
+
+function startDeflateWorkerSingleton(createDeflateWorkerImpl: typeof createDeflateWorker) {
+  try {
```

Comment:
It could be interesting to add a comment on why they are a `try catch` and an `error event listener`. (Chromium/Firefox)

Reply:
it could be nice to check if there is some spec on the behavior and see if an issue should be opened somewhere

Reply:
I've opened a bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1736865